src/document/document.ts

```diff
@@ -660,7 +660,13 @@ export class Document<T, P extends Indexable = Indexable> {
     } catch (err) {
       // drop clone because it is contaminated.
       this.clone = undefined;
-      logger.error(err);
+
+      if (err instanceof YorkieError) {
+        logger.error(err.toString());
+      } else {
+        logger.error(err);
+      }
+
       throw err;
```

Comment:
Is it okay if we don't consistently throw `YorkieError` here? 
For example, if we test by intentionally causing an error in the [public/index.html](https://github.com/yorkie-team/yorkie-js-sdk/blob/83a6369f11de43d59d457302283c8ac1c89bdb23/public/index.html#L341) example with `doc.update((root) => {root.a.b = 1;})`, it outputs a general error message.

![image](https://github.com/user-attachments/assets/2fd4bec7-a4b6-46a6-8f1d-4ea32911c662)

Reply:
Thank you for pointing this out.

In the example you suggested, it seems like there is an intentional error within the `updater`. As expected, the error will be thrown with `throw err` back to the caller in `public/index.html`. When the error is caught in the caller, it is logged using the default `console.error`, which results in the behavior as screenshot above.

```typescript
// document.ts
 try {
      // skipped
      updater(
        proxy,
        new Presence(context, this.clone!.presences.get(actorID)!),
      );
    } catch (err) {
      // skipped
      if (err instanceof YorkieError) {
        logger.error(err.toString());
      } else {
        logger.error(err);
      }

      throw err;
    }
```

```typescript
// public/index.html
async function main(){
   try{
     // intentional error
     doc.update((root) => {root.a.b = 1;})
   }catch(e){
      console.error(e)
   }
}
```

To resolve this, I believe two things need to be done.

> 1.  Replace `console.error` with `logger.error`.

When logging with the default logger such as `console.error`, even if a `YorkieError` is thrown, the error message won't be customized to start with `YORKIE E:`. Therefore, replacing `console.error` with our custom `logger` seems necessary to ensure consistent format of error messages.

> 2. throw `YorkieError` instead of default `err`

Instead of `throw err`, I think it's a great idea to throw a `YorkieError`. For example, it would look like as follows: 

```javascript
// document.ts
 try {
      // skipped
    } catch (err) {
       // skipped
       throw new YorkieError(Code.ErrSomething, "failed to execute updater").
    }
```

However, I'm concerned that it is difficult to determine the exact cause of the error, which makes me hesitant to use `YorkieError`. To be specific, in this case there could be an error in `updater` or just another function within `try` phrase like `createJSON`. Therefore, I doubt that it is a good idea to `throw new YorkieError(Code.ErrSomething, "failed to execute updater").` 

Of course, we could address this issue by using a more general error message, such as `throw new YorkieError(Code.ErrSomething, "failed to update document")`. However, this approach would make it harder to debug the issue in the future.

Reply:
Moreover, I'm concerned that in the example you showed, I believe there will be two error messages logged with the same content. One logged with `logger.error` and the other with `console.error`. Just to clarify:

```typescript
// document.ts
try{
  updater(
        proxy,
        new Presence(context, this.clone!.presences.get(actorID)!),
      );
}
 catch (err) {
      if (err instanceof YorkieError) {
        logger.error(err.toString());
      } else {
        // logs first time
        logger.error(err);
      }

      throw err
}
```

```typescript
// public/index.html

async function main(){
   try{
     doc.update((root) => {root.a.b = 1;})
   }catch(e){
     // logs second time
      console.error(e)
   }
}
```

Perhaps there is a way to remove one of the logger calls from either the `document` or `public/index.html`, but I'm not sure it's the best way. Please let me know your thoughts on this.

Additionally, I would like to reproduce the error in the snapshot but I'm not sure how to do it. Did you insert the code directly or test it through the console? Any guidance on reproducing the error would be appreciated.

Reply:
@gwbaik9717 

I added `root.a.b = 1;` to line 345 of `index.html`. 
As you mentioned, this causes the log to be printed twice.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/ce04de43-8ae7-448b-ab6e-364d06554bcb">
<img width="300" alt="image" src="https://github.com/user-attachments/assets/ea88a32d-8d1a-4bb2-a380-c02370bf6ded">

If we set `yorkie.setLogLevel(yorkie.LogLevel.Fatal);` on [line 319](https://github.com/yorkie-team/yorkie-js-sdk/blob/83a6369f11de43d59d457302283c8ac1c89bdb23/public/index.html#L319), the log is printed only once.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/f26240bb-8a0c-4737-90e6-dbcbcf177fda">

Therefore, the visibility of the first log is determined by the logger level set by the user.

Reply:
As you suggested, I tested wrapping general errors with `YorkieError`, but there doesn't seem to be a significant difference in usability. 

```js
// document.ts
if (err instanceof YorkieError) {
  throw err;
} else {
  throw new YorkieError(Code.ErrDummy, err.toString());
}
```

<img width="1236" alt="image" src="https://github.com/user-attachments/assets/6a3f7086-0da8-4bfa-80eb-4f2eac9f8251">


I initially thought it would be clearer because `YorkieError:` is displayed at the beginning of the error(`[2-1]`), but even when throwing errors normally, we can see from the call stack that the error occurred in `yorkie-js-sdk`.(refer to the blue highlighted section in image `[2-2]`) So it seems like it might be okay not to necessarily wrap errors with YorkieError. Thanks.

( BTW, I discovered that the `message` argument is not being used in the `fatal` method. It's good to remove it. )

<img width="400" alt="image" src="https://github.com/user-attachments/assets/dff6e0ea-ae6a-4061-8599-fc8e4f87c06d">